### PR TITLE
Disallow modifications to `pybamm.Simulation` object attributes after initialisation

### LINKED ISSUE #3196: pybamm.Simulation - private attributes ignore current state of public attributes in build() and solve()
In the `pybamm.Simulation` class, some behaviour of implicitly public attributes is ambiguous / poorly documented.

In particular, the `build()` and `solve()` functions assume that the values of `self.parameter_values` and `self.model` equal those in the initial state of the `pybamm.Simulation` instance following `__init__()`, since they act on private attributes that are set to the initial state of theser variables. This conflicts with the API's implication that these attributes are user-editable.

Example:

```
model = pybamm.lithium_ion.DFN()
# User wants to discard automatically generated events
model.events = []
sim = pybamm.Simulation(
   model,
   pybamm.ParameterValues("Chen2020")
)
sim.solve()    # Runs with no events
```

whereas

```
model = pybamm.lithium_ion.DFN()
sim = pybamm.Simulation(
   model,
   pybamm.ParameterValues("Chen2020")
)
# User wants to discard automatically generated events
sim.model.events = []
sim.solve()     # Runs with events anyway!
```

This is because `solve()` replaces `self.model` using a call to `set_parameters()` with `sim._unprocessed_model`, which preserves the state of `model` when `sim` was initialised.

Other affected attributes:

* `self._unprocessed_parameter_values` - preserves initial state of `self.parameter_values`, irrespective of any subsequent changes
* `self.operating_mode` - determined based on the initial condition, irrespective of any subsequent changes

Better API behaviour:

* Re-initialise private attributes based on the public state of the at time of use, not at time of initialisation.
* Or, if direct edits to `sim.parameter_values` and `sim.model` are strictly not allowed after `sim` is initialised, they should be declared private (e.g. `sim._model`).

In any case, the API documentation could be clearer about the activities of `set_parameters()` and `build()`, and when direct external calls are required to these functions.

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
# Description

Removes the setter methods in the `Simulation` class, so that parameters defined in the initialisation of a `pybamm.Simulation` instance cannot be modified unless the instance is redefined

This way, attributes like `sim.model` or `sim.parameter_values` cannot be edited directly. A new `sim` object with the modified parameters is required to incorporate changes, i.e.,

```python
import pybamm

model = pybamm.lithium_ion.DFN()
sim = pybamm.Simulation(model)

sim.model = pybamm.lithium_ion.DFN()  # no setter method for self.model(), will not work

model1 = pybamm.lithium_ion.SPM()
sim1 = pybamm.Simulation(model)  # new simulation object with the SPM, works
```

Fixes #3196

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
@tinosulzer let me know if this needs a CHANGELOG entry (my guess is that it does)

**Review by brosaplanella — COMMENTED** [review]
Looks good, thanks! Can you improve the coverage and add a CHANGELOG line? Happy to approve then :)

**Comment by agriyakhetarpal** [comment]
I added a CHANGELOG entry. Could you help on improving the coverage? It passes locally for me

**Comment by brosaplanella** [comment]
> I added a CHANGELOG entry. Could you help on improving the coverage? It passes locally for me

Ok, then probably it is an issue with Codecov. Let's wait for a second review before merging in any case.

File: CHANGELOG.md
Comment (brosaplanella) [inline CHANGELOG.md:12]: Can you sort the entries in this section so they are in decreasing PR number?

Comment (agriyakhetarpal) [reply CHANGELOG.md]: Done

**Review by brosaplanella — APPROVED** [review]
Looks good, thanks!

**Pull request merged**
